File: api/algorithms/kmeans_1d.py
import numpy as np

from api.algorithms._shared import (
    compute_histogram,
    palette_from_zones,
    zones_from_boundaries,
)


def _kmeans_plus_plus_init(
    values: np.ndarray, weights: np.ndarray, k: int, rng: np.random.Generator
) -> np.ndarray:
    # Weighted k-means++: first center sampled proportional to weight; subsequent
    # centers sampled proportional to weight * (min squared distance to chosen centers).
    probs = weights / weights.sum()
    first = int(rng.choice(len(values), p=probs))
    centers = [values[first]]
    min_d2 = (values - centers[0]) ** 2
    for _ in range(1, k):
        p = weights * min_d2
        s = p.sum()
        idx = int(rng.choice(len(values), p=(p / s) if s > 0 else probs))
        centers.append(values[idx])
        min_d2 = np.minimum(min_d2, (values - values[idx]) ** 2)
    return np.array(sorted(centers), dtype=float)


def _lloyd_1d(
    values: np.ndarray,
    weights: np.ndarray,
    centers: np.ndarray,
    max_iter: int = 100,
    tol: float = 1e-6,
) -> np.ndarray:
    for _ in range(max_iter):
        midpoints = (centers[:-1] + centers[1:]) / 2
        assignments = np.digitize(values, midpoints)
        new_centers = np.empty_like(centers)
        for k in range(len(centers)):
            mask = assignments == k
            w = weights[mask]
            if w.sum() == 0:
                new_centers[k] = centers[k]
            else:
                new_centers[k] = (values[mask] * w).sum() / w.sum()
        new_centers.sort()
        if np.max(np.abs(new_centers - centers)) < tol:
            return new_centers
        centers = new_centers
    return centers


def run(gray: np.ndarray, n: int = 5, seed: int = 0):
    counts = compute_histogram(gray).astype(float)
    values = np.arange(256, dtype=float)
    mask = counts > 0
    v_used = values[mask]
    w_used = counts[mask]
    k = min(n, len(v_used))
    if k < 2:
        raise ValueError("not enough distinct luminosity values")
    rng = np.random.default_rng(seed)
    centers = _kmeans_plus_plus_init(v_used, w_used, k, rng)
    centers = _lloyd_1d(v_used, w_used, centers)
    centers.sort()
    boundaries = np.round((centers[:-1] + centers[1:]) / 2).astype(int)
    zones = zones_from_boundaries(gray, boundaries)
    palette = palette_from_zones(gray, zones)
    return zones, palette, boundaries

File: frontend/app/lib/algorithms/kmeans1d.ts
// 1-D weighted k-means over a 256-bin grayscale histogram. Ported from
// api/algorithms/kmeans_1d.py. Operates on values 0..255 with their
// histogram counts as weights, so cost is independent of image size
// once the histogram is computed.
//
// RNG note: numpy's default_rng uses PCG64; we use mulberry32 here for
// determinism with a tiny footprint. Same algorithm, same convergence
// behavior — but seeded outputs will not byte-match the Python pipeline.

function mulberry32(seed: number): () => number {
  let t = seed >>> 0;
  return () => {
    t = (t + 0x6d2b79f5) >>> 0;
    let r = t;
    r = Math.imul(r ^ (r >>> 15), r | 1);
    r ^= r + Math.imul(r ^ (r >>> 7), r | 61);
    return ((r ^ (r >>> 14)) >>> 0) / 4294967296;
  };
}

function weightedChoice(
  weights: Float64Array,
  total: number,
  rand: () => number,
): number {
  // Inverse-CDF sample. Linear scan is fine: weights.length <= 256.
  let r = rand() * total;
  for (let i = 0; i < weights.length; i++) {
    r -= weights[i];
    if (r <= 0) return i;
  }
  return weights.length - 1;
}

function kmeansPlusPlusInit(
  values: Float64Array,
  weights: Float64Array,
  k: number,
  rand: () => number,
): Float64Array {
  const n = values.length;
  let weightSum = 0;
  for (let i = 0; i < n; i++) weightSum += weights[i];

  const first = weightedChoice(weights, weightSum, rand);
  const centers: number[] = [values[first]];

  const minD2 = new Float64Array(n);
  for (let i = 0; i < n; i++) {
    const d = values[i] - centers[0];
    minD2[i] = d * d;
  }

  const p = new Float64Array(n);
  for (let c = 1; c < k; c++) {
    let s = 0;
    for (let i = 0; i < n; i++) {
      p[i] = weights[i] * minD2[i];
      s += p[i];
    }
    const idx =
      s > 0 ? weightedChoice(p, s, rand) : weightedChoice(weights, weightSum, rand);
    centers.push(values[idx]);
    for (let i = 0; i < n; i++) {
      const d = values[i] - values[idx];
      const d2 = d * d;
      if (d2 < minD2[i]) minD2[i] = d2;
    }
  }

  centers.sort((a, b) => a - b);
  return Float64Array.from(centers);
}

function lloyd1D(
  values: Float64Array,
  weights: Float64Array,
  centers: Float64Array,
  maxIter = 100,
  tol = 1e-6,
): Float64Array {
  const n = values.length;
  const k = centers.length;
  const cur = Float64Array.from(centers);
  const next = new Float64Array(k);
  const sums = new Float64Array(k);
  const ws = new Float64Array(k);

  for (let iter = 0; iter < maxIter; iter++) {
    sums.fill(0);
    ws.fill(0);

    // Assign each value to its nearest center using midpoint thresholds.
    // Single linear sweep: values are sorted (0..255 ascending).
    let kIdx = 0;
    for (let i = 0; i < n; i++) {
      const v = values[i];
      while (kIdx < k - 1 && v >= (cur[kIdx] + cur[kIdx + 1]) / 2) kIdx++;
      const w = weights[i];
      sums[kIdx] += v * w;
      ws[kIdx] += w;
    }

    for (let c = 0; c < k; c++) {
      next[c] = ws[c] === 0 ? cur[c] : sums[c] / ws[c];
    }

    // Sort in place — k is small.
    Array.prototype.sort.call(next, (a: number, b: number) => a - b);

    let maxDelta = 0;
    for (let c = 0; c < k; c++) {
      const d = Math.abs(next[c] - cur[c]);
      if (d > maxDelta) maxDelta = d;
      cur[c] = next[c];
    }
    if (maxDelta < tol) return cur;
  }
  return cur;
}

export function runKmeans1D(
  histogram: Uint32Array,
  n: number,
  seed = 0,
): { boundaries: Int32Array; zoneCount: number } {
  // Restrict to non-zero bins (matches the Python "mask = counts > 0").
  const usedValues: number[] = [];
  const usedWeights: number[] = [];
  for (let i = 0; i < 256; i++) {
    if (histogram[i] > 0) {
      usedValues.push(i);
      usedWeights.push(histogram[i]);
    }
  }
  const values = Float64Array.from(usedValues);
  const weights = Float64Array.from(usedWeights);
  const k = Math.min(n, values.length);

  const rand = mulberry32(seed);
  let centers = kmeansPlusPlusInit(values, weights, k, rand);
  centers = lloyd1D(values, weights, centers);
  Array.prototype.sort.call(centers, (a: number, b: number) => a - b);

  const boundaries = new Int32Array(Math.max(0, k - 1));
  for (let i = 0; i < boundaries.length; i++) {
    boundaries[i] = Math.round((centers[i] + centers[i + 1]) / 2);
  }
  return { boundaries, zoneCount: k };
}
